reservations/allocation_solver.py

```diff
@@ -72,38 +86,44 @@ def print_solution(self):
 
 class AllocationSolver(object):
     def __init__(self, allocation_data: AllocationData):
-        self.spaces = allocation_data.spaces
+        self.spaces: Dict[int, AllocationSpace] = allocation_data.spaces
         self.allocation_events = allocation_data.allocation_events
 
     def solve(self):
         model = cp_model.CpModel()
 
         selected = {}
-        for space in self.spaces:
-            for allocation_event in self.allocation_events:
-                for occurence_id, occurence in allocation_event.occurrences.items():
+        for allocation_event in self.allocation_events:
+            for occurence_id, occurence in allocation_event.occurrences.items():
+                for space_id, space in suitable_spaces_for_event(
```

Comment:
This space restriction could be also done with constraints on the model, while this might be a bit less readable, i think it might be smart to do it this way instead so we limit the amount of data we pass into the solver as early as possible to improve performance. This would be a hard constraint in the model anyway, so doing it this way won't affect the solutions.